src/python/pants/backend/go/tailor.py

```diff
@@ -0,0 +1,147 @@
+    for dirname, filenames in group_by_dir(unowned_go_files).items():
+        putative_targets.append(
+            PutativeTarget.for_target_type(
+                GoPackage,
+                dirname,
+                os.path.basename(dirname),
+                sorted(filenames),
+            )
+        )
+
+    return PutativeTargets(putative_targets)
+
+
+@dataclass(frozen=True)
+class PutativeGoModuleTargetsRequest(PutativeTargetsRequest):
+    pass
+
+
+@rule(level=LogLevel.DEBUG, desc="Determine candidate Go `go_module` targets to create")
+async def find_putative_go_module_targets(
+    request: PutativeGoModuleTargetsRequest, all_owned_sources: AllOwnedSources
+) -> PutativeTargets:
+    all_go_mod_files = await Get(Paths, PathGlobs, request.search_paths.path_globs("go.mod"))
+    unowned_go_mod_files = set(all_go_mod_files.files) - set(all_owned_sources)
+
+    putative_targets = []
+    for dirname, filenames in group_by_dir(unowned_go_mod_files).items():
+        putative_targets.append(
+            PutativeTarget.for_target_type(
+                GoModule,
+                dirname,
+                os.path.basename(dirname),
+                sorted(filenames),
+            )
+        )
+
+    return PutativeTargets(putative_targets)
+
+
+def compute_go_external_module_target_name(name: str, version: str) -> str:
+    return f"{name.replace('/', '_')}_{version}"
+
+
+@dataclass(frozen=True)
+class PutativeGoExternalModuleTargetsRequest(PutativeTargetsRequest):
+    pass
+
+
+@rule(level=LogLevel.DEBUG, desc="Determine candidate Go `go_external_module` targets to create")
+async def find_putative_go_external_module_targets(
+    request: PutativeGoExternalModuleTargetsRequest, _all_owned_sources: AllOwnedSources
+) -> PutativeTargets:
+    # Unlike ordinary tailor invocations, this rule looks at existing `go_module` targets and not at actual
+    # source files because it infers `go_external_module` targets based on go.mod contents. (This may require
+    # invoking `tailor` first to create `go_module` targets and then again to create `go_external_module`
+    # targets.)
+    #
+    # TODO: This might better work as a BUILD macro if https://github.com/pantsbuild/pants/issues/7022 is
+    # resolved and macros are able to invoke the engine or processes.
+
+    addresses = itertools.chain.from_iterable(
+        [
+            [MaybeEmptySiblingAddresses(search_path), MaybeEmptyDescendantAddresses(search_path)]
+            for search_path in request.search_paths.dirs
+        ]
+    )
+    candidate_targets = await Get(UnexpandedTargets, AddressSpecs(addresses))
+    go_module_targets = [
+        tgt
+        for tgt in candidate_targets
+        if tgt.has_field(GoModuleSources) and not tgt.address.is_file_target
+    ]
+
+    putative_targets = []
+
+    for go_module_target in go_module_targets:
+        resolved_go_module = await Get(
+            ResolvedGoModule, ResolveGoModuleRequest(go_module_target.address)
+        )
```

Comment:
Please either switch this to use `MultiGet` or add a TODO to fix it. You get worse perf using `Get` in a for loop because it's sequential.

Reply:
(You can add the TODO in a followup if you're touching this code soon in a followup)